### PR TITLE
node: strain npm installation

### PR DESCRIPTION
**Pull request by xu-cheng**
I think we should establish a policy that formulae should never be
allowed to write outside `Cellar`, `etc`, `share` and `lib`.

Users are required to run `brew link --overwrite node` to migrate from
the original installation to this one. But this is the direct result of
incorrect old behaviors.

This can also help me to sandbox `postinstall`. :smile:

### CODE REVIEW

**Comment by mikemcquaid** [comment]
Can you explain a bit about what the change is here and the motivation for it? I'm finding it hard to follow.

> I think we should establish a policy that formulae should never be
> allowed to write outside Cellar, etc, share and lib.

I agree with this generally but with e.g. `npm` it's more important that the installation is consistent with how `npm` is installed by upstream than it meeting our conventions.

**Comment by xu-cheng** [comment]
Basically two changes made in this PR.
* Instead of symlinking from `HOMEBREW_PREFIX/"lib/node_modules/npm/bin/npm-cli.js"` to `HOMEBREW_PREFIX/"bin/npm"` manually. I create a shim `prefix/"bin/npm"` which points to `HOMEBREW_PREFIX/"lib/node_modules/npm/bin/npm-cli.js"`. And let Homebrew symlink it to `HOMEBREW_PREFIX` automatically. 
* Instead of symlinking manpage from `libexec/"npm/share/man"` to `HOMEBREW_PREFIX` manually. I create symlinks from `libexec/"npm/share/man"` to `prefix/"share/man"`. Homebrew will handle rest of it.

In conclusion, a formula should not be allowed link itself from Cellar to prefix by itself. This should be done by Homebrew.

**Comment by mikemcquaid** [comment]
> Instead of symlinking from HOMEBREW_PREFIX/"lib/node_modules/npm/bin/npm-cli.js" to HOMEBREW_PREFIX/"bin/npm" manually. I create a shim prefix/"bin/npm" which points to HOMEBREW_PREFIX/"lib/node_modules/npm/bin/npm-cli.js". And let Homebrew symlink it to HOMEBREW_PREFIX automatically.

My concern is that this results in `npm` putting files or considering itself to be in a different path. If you can show this isn't the case: :cool:

> Instead of symlinking manpage from libexec/"npm/share/man" to HOMEBREW_PREFIX manually. I create symlinks from libexec/"npm/share/man" to prefix/"share/man". Homebrew will handle rest of it.

My concern here is that we're using `ln_sf` for a reason here so we can overwrite existing `npm` installs. I'm concerned if someone has installed `npm` before trying to do so with Homebrew this cause `brew link` to fail.

I'd like @DomT4 to give some input/thoughts here.

**Comment by xu-cheng** [comment]
> My concern is that this results in npm putting files or considering itself to be in a different path. If you can show this isn't the case: :cool:

There's a little trouble in `npm` upgrade. It will overwrite the symlink which will cause `brew link` to fail. I will look into it. Hope to find some workaround. :stuck_out_tongue_winking_eye: 

> My concern here is that we're using ln_sf for a reason here so we can overwrite existing npm installs. I'm concerned if someone has installed npm before trying to do so with Homebrew this cause brew link to fail.

In fact, I see this as a normal result. It's also consistent with other formulae. e.g. a user install pip using system python will get similar result when he/she tries to install brew python. On the contrary, current behaviors have a few problems. `brew unlink node` didn't affect npm and `brew rm node` would leave some residual symlinks.

**Comment by mikemcquaid** [comment]
> There's a little trouble in npm upgrade. It will overwrite the symlink which will cause brew link to fail. I will look into it. Hope to find some workaround.

Yup, this is one of the reasons it's done the way it is.

> In fact, I see this as a normal result. It's also consistent with other formulae. e.g. a user install pip using system python will get similar result when he/she tries to install brew python. On the contrary, current behaviors have a few problems. brew unlink node didn't affect npm and brew rm node would leave some residual symlinks.

It's also pretty user unfriendly. I think it's OK for this stuff to be done in `postinstall` just not in `install`.

**Comment by xu-cheng** [comment]
If I cannot find a better solution, I will have to put `/bin/ln` outside of sandbox when I implement sandbox `postinstall`. To be honest, I don't quite in favor of this.

**Comment by mikemcquaid** [comment]
Let's step back and ask what the sandbox is for: is it to restrict what formula authors can do or is it to protect users from potentially malicious code from packaged software. I personally believe and think it is for the latter and that the prior is more likely to be annoying than useful, particularly for third-party and tapped formulae. I'm wondering whether, as a result, we even want to sandbox `postinstall` (although `test` still makes sense to me).

**Comment by xu-cheng** [comment]
Yeah, you are right. In the end, it comes to one question in what level of security we want to achieve? We need to find a balance between security and usability. So I will pass this PR.

For implementing sandbox `postinstall`, I think I have two choices: a) allowing write access to whole HOMEBREW_PREFIX for all formulae. b) only allowing write access to whole HOMEBREW_PREFIX for selective formulae.  i.e. create a some kind whitelist. What's your opinion?

**Comment by DomT4** [comment]
> I create symlinks from libexec/"npm/share/man" to prefix/"share/man". Homebrew will handle rest of it.

Sadly last time we did this, we had failures galore. Because of the way we historically installed ` npm ` (i.e. directly into HOMEBREW_PREFIX) it meant Homebrew wasn't so great at recognising it'd actually put the manpages there in the first place and would refuse to override real manpages with symlinks. The somewhat brutal ` rm_f ` and ` ln_sf ` saved us on this.

Given the amount of time that has passed since we implemented that, we could look at changing it, but the whole Node ecosystem has a lot of variables that make life for us difficult.

Essentially the ` npm ` install method here duplicates almost step-for-step what the Node ecosystem itself does. Obviously we're installing into ` libexec ` and then moving specific things, but the methodology itself is as close to identical as we can get.

> On the contrary, current behaviors have a few problems.

This is largely because of the steps we took. If you look back 4/5 Node versions ago you should find the issue where we had to fix the man placement and the discussions there :).

>  It has to be allowed write access to whole HOMEBREW_PREFIX.

Yeh. If you deny it native access to ` bin `, `include`, ` lib ` or ` share ` it *should* tantrum. It needs all of those things for ` npm ` modules. Theoretically, include could be ` opt_include `.

Those are my general thoughts. The proposal isn't bad, but I'm not sure what we gain from it compared to the risk of upsetting the delicate Node-balance we have at the moment :smile_cat:. It's hard to sandbox something like ` npm ` due to the sheer volume of filesystem and network connections it makes to install modules (Including, sometimes, the $HOME directory).

**Comment by xu-cheng** [comment]
>  It's hard to sandbox something like npm due to the sheer volume of filesystem and network connections it makes to install modules (Including, sometimes, the $HOME directory).

@DomT4 a little note here. When we talk about sandbox, we only mean sandboxing `brew install`, `brew postinstall`, etc. We won't do anything after it left homebrew space.

**Comment by DomT4** [comment]
> When we talk about sandbox, we only mean sandboxing

Yeah, I know. I'm already running your sandbox locally, heh. I just meant in general it's hard to sandbox things like that. I killed the comment just so clarity wasn't accidentally fuzzed there. Apologies for the confusion :smiley_cat:

**Comment by xu-cheng** [comment]
No problem. And I'm glad that you hope to test the sandbox. :bow:

**Comment by mikemcquaid** [comment]
> For implementing sandbox postinstall, I think I have two choices: a) allowing write access to whole HOMEBREW_PREFIX for all formulae. b) only allowing write access to whole HOMEBREW_PREFIX for selective formulae. i.e. create a some kind whitelist. What's your opinion?

I'd strongly prefer a).
